### ISSUE TITLE
Support multiple tenant connections per user

### ISSUE THREAD

**Issue by ivorscott** [devpies/saas-core#38]
Related to #35 

## Problem

Multiple tenant connections per user isn't supported. Users should be able to switch between tenants accounts in the UI if they have access.

**Comment by ivorscott** [comment]
Initial thoughts: 
1) How do users switch between tenants? 
2) How do users switch between tenants on different plans? 
3) Where does login fit in?
<img width="1628" alt="Screen Shot 2022-06-28 at 03 59 14" src="https://user-images.githubusercontent.com/1488558/176072716-1ffe17ef-e7aa-4f75-b5ee-8e56719e8f4b.png">

**Comment by ivorscott** [comment]
## Strategy

Modify tokens with tenant connections before delivering them to users. The id token will still have a `custom:tenant-id` attribute (which is the first tenant account associated with the user), in addition to the `custom:tenant-connections attribute`. The tenant context flows through the app services like before but now there are multiple tenant connections. A backend service will derive the "active" tenant for a request by checking the http referrer (e.g., "devpie.io/tenanta" or "devpie.io/tenantb"). The path is the company name formatted into a unique identifier called the tenant path. The tenant path is also stored as a JSON string array in custom:tenant-connections.

<img width="650" alt="Screen Shot 2022-07-04 at 02 38 13" src="https://user-images.githubusercontent.com/1488558/177063332-ecd016b1-ab26-4da2-899a-ba9ca055cc31.png">

### How it works

A custom AWS Lambda function is invoked by Cognito in response to a PreTokenGeneration trigger. The event is `events.CognitoEventUserPoolsPreTokenGen` in Go. So upon sign in, or whenever a token refresh is requested, a Lambda function fires, and it adds tenant connections as a custom attribute before the idToken is actually generated.

https://github.com/aws/aws-lambda-go/blob/main/events/README_Cognito_UserPools_PreTokenGen.md
https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html

```go
func handler(ctx context.Context, event events.CognitoEventUserPoolsPreTokenGen) (events.CognitoEventUserPoolsPreTokenGen, error) {
	var err error

	client := NewDynamoRepository(ctx, event.Region)

	tenants, err := client.FindTenants(ctx, event.Request.UserAttributes["sub"])
	if err != nil {
		return event, err
	}
	fmt.Printf("Tenants: %+v \n", tenants)

	connections := newTenantConnections(tenants)
	fmt.Printf("Tenants Connections: %+v \n", connections)

	bytes, err := json.Marshal(&connections)
	if err != nil {
		return event, err
	}

	event.Response.ClaimsOverrideDetails.ClaimsToAddOrOverride = map[string]string{
		"tenant-connections": string(bytes),
	}
	return event, nil
}
```

### Token Refresh with Websockets

If a user is already logged in and a tenant connection is added to them, refresh their token on the fronted in the background. We can do this with push notifications via websockets. 

#### Scenario: User is already logged in

<img width="1399" alt="Screen Shot 2022-07-04 at 02 59 35" src="https://user-images.githubusercontent.com/1488558/177064506-64e87aeb-a78c-460a-89e9-26107a322f5e.png">

1. Create AWS Lambda function to create tenant context.
2. Connect Cognito PreTokenGeneration trigger to Lambda function.
3. Frontend requests the user service to add a user to tenant.
4. User is added to pool if they don't exist.
5. User is added to database if they don't exist.
6. Tenant to user mapping is added to the database, not the token by default.
7. Backend notifies frontend to refresh token via websocket.
8. Refresh token .
9. PreTokenGeneration hook Invokes Lambda function.
9. Frontend gets new token with tenant context.
10. Sidebar renders tenant connections.

### Shared User Pool VS Dedicated User Pool 

I had planned having dedicated user pools for premium tenants but that approach has a cost:
1. Multiple logins. Users with tenant connection from mixed plans(on basic and premium) would need to login again to access another tenant connection.
2. Confusing user experience. The app shows all tenants connected to the user in the sidebar. Navigating between tenants you have access to is hindered by potentially needing to login more than once.
3. Routing is complex. Dedicated user pools need a dedicated app front-ends at reserved locations (e.g., subdomain.devpie.io/tenantc) to wrap auth middleware around. Therefore, the url will change from devpie.io and subdomain.devpie.io which is weird. AWS amplify does support overriding the active userpool used for auth dynamically but this approach is complex, hard to maintain and wouldn't improve the user experience.

#### Solution: A Single Shared User Pool

It's much easier to use a single shared user pool for all application users. This creates a unified experience because users sign in once. For the front-end to determine which backends to send requests to, the "active" tenant must be derived from a routing scheme: `devpie.io/tenant_path`. Then the api client on the frontend needs to check the tenant plan and map requests to the correct api endpoint. If the plan is basic: the api endpoint is `/api/*`. If the plan is premium: the api endpoint is `/TENANT_PATH/api/*`.  The ingress controller will map requests to the right ingress resources in the desired cluster namespaces: either the default namespace with shared app services or dedicated resources in a tenant specific namespace.

<img width="1489" alt="Screen Shot 2022-07-04 at 03 28 39" src="https://user-images.githubusercontent.com/1488558/177066319-b4d8a511-e104-4428-b770-aec4ede324a9.png">

**Comment by ivorscott** [comment]
This feature may be much easier than I anticipated. While improving the ModifyTokenFunc AWS Lambda function, I realized the app receives token changes automatically. So there's no need to log out and in again or push a notification from the server side. This is great news. AWS Amplify probably refreshes the token automatically.

The new version of the ModifyTokenFunc replaces the tenant connection JSON array with a map indexed by tenantId to improve time complexity.

<img width="1841" alt="Screen Shot 2022-07-08 at 00 11 20" src="https://user-images.githubusercontent.com/1488558/177880336-80850168-f576-4764-8651-3f2da42fe9df.png">
